### PR TITLE
Categories

### PR DESCRIPTION
**Pull request by jroodman**
This PR adds full grocery categorization functionality to the Pantry application (expanding on the dairy categorization functionality in the previous PR).

### CODE REVIEW

File: lib/pantry/helpers/category_helper.rb [hunk 1]
```diff
@@ -56,14 +57,15 @@ def determine_large_category(item_name)
...
+            :dairy          => Pantry::Helpers::DairyCategoryHelper,
+            :meat           => Pantry::Helpers::MeatCategoryHelper,
+            :poultry        => Pantry::Helpers::PoultryCategoryHelper,
+            :eggs           => Pantry::Helpers::EggCategoryHelper,
+            :seafood        => Pantry::Helpers::SeafoodCategoryHelper,
+            :fruit          => Pantry::Helpers::FruitCategoryHelper,
+            :vegetable      => Pantry::Helpers::VegetableCategoryHelper,
+            :grain          => Pantry::Helpers::GrainCategoryHelper,
+            :other          => Pantry::Helpers::OtherCategoryHelper
           }[large_category]
```
Comment (efatsi) [inline lib/pantry/helpers/category_helper.rb]: completely up to you, but you could probably pull this off with some fhancy meta programming:
```ruby
def helper_for(category)
  "Pantry::Helpers#{category.classify}CategoryHelper".constantize
end
```
benefit would be if you added new categories you wouldn't have to mess with this method. downside is that it's a little cryptic, and it looks like `:eggs` would have to be handled individually or renamed.

**Comment by efatsi** [comment]
cool stuff, a couple things that stick out:
- in all of your sub category methods, you're calling `words = ...`. Unless I'm missing something, I don't think that variable is serving any purpose and could probably be removed
- you're using `Set.new` a lot, are you using some specific functionality from the Set class to warrant that? If not, I think you could just stick with defining the raw arrays, or `[]` for the empty array.

File: lib/pantry/helpers/dairy_category_helper.rb [hunk 2]
```diff
@@ -39,32 +39,8 @@ class DairyCategoryHelper
...
-          CATEGORIES.reduce(Set.new) do |set, (k,v)|
-            set.merge method(k).to_proc.call
-          end
-        end
-
-        private
-
         def milk
-          products = Set.new([
+          words = Set.new([
```
Comment (albaer) [inline lib/pantry/helpers/dairy_category_helper.rb]: I don't think you need to set this to a variable in any of these methods.

File: lib/pantry/helpers/subcategory_helper.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        def time_til_warning(category:)
+          self::CATEGORIES[category][:warning]
+        end
+
+        def time_til_expiration(category:)
+          self::CATEGORIES[category][:expiring]
+        end
+
+        def allWords
```
Comment (albaer) [inline lib/pantry/helpers/subcategory_helper.rb]: snake_case_for_ruby (unless I'm missing something)

**Review by albaer — COMMENTED** [review]
Cool stuff! I feel like I could use this. Just a couple of very small comments.

File: lib/pantry/helpers/subcategory_helper.rb (hunk 3 above)
Comment (jroodman) [reply lib/pantry/helpers/subcategory_helper.rb]: Yep! Will change that. My Java heritage is still looming over me.

File: lib/pantry/helpers/dairy_category_helper.rb (hunk 2 above)
Comment (jroodman) [reply lib/pantry/helpers/dairy_category_helper.rb]: Indeed... thanks!

**Comment by jroodman** [comment]
@efatsi 
I chose to use sets due to the efficiency of checking if they contain a specific value. Arrays I believe max out at n if they are not sorted. Sets should ideally be log(n) or faster. 

As well, all of the sets are combined in the all_words function, using sets is an easy way to deal with duplicates.

File: lib/pantry/helpers/category_helper.rb (hunk 1 above)
Comment (jroodman) [reply lib/pantry/helpers/category_helper.rb]: I like this... also very shmancy.

**Pull request merged**